tracing-subscriber/src/fmt/format/json.rs

```diff
@@ -89,37 +120,21 @@ where
             let id = id.id();
             if let Some(id) = id {
                 if let Some(span) = ctx.ctx.span(id) {
-                    let ext = span.extensions();
-                    let data = ext
-                        .get::<FormattedFields<N>>()
-                        .expect("Unable to find FormattedFields in extensions; this is a bug");
-                    // TODO: let's _not_ do this, but this resolves
-                    // https://github.com/tokio-rs/tracing/issues/391.
-                    // We should probably rework this to use a `serde_json::Value` or something
-                    // similar in a JSON-specific layer, but I'd (david)
-                    // rather have a uglier fix now rather than shipping broken JSON.
-                    let mut fields = match serde_json::from_str(&data) {
-                        Ok(fields) => fields,
-                        // We have previously recorded fields for this span
-                        // should be valid JSON. However, they appear to *not*
-                        // be valid JSON. This is almost certainly a bug, so
-                        // panic if we're in debug mode
-                        Err(e) if cfg!(debug_assertions) => panic!(
-                            "span '{}' had malformed fields! this is a bug.\n  error: {}\n  fields: {:?}",
-                            span.metadata().name(),
-                            e,
-                            data
-                        ),
-                        // If we *aren't* in debug mode, it's probably best not
-                        // crash the program, but let's at least make sure it's clear
-                        // that the fields are not supposed to be missing.
-                        Err(e) => json!({ "field_error": format!("{}", e) }),
-                    };
-                    fields["name"] = json!(span.metadata().name());
-                    serializer.serialize_entry("span", &fields).unwrap_or(());
+                    serializer
+                        .serialize_entry("span", &visit_span(&span))
+                        .unwrap_or(());
                 }
             }
 
+            let mut spans = vec![];
+            ctx.visit_spans(|span| {
+                spans.push(visit_span(span));
+                Ok(())
+            })?;
```

Comment:
take it or leave it: it might be a bit more idiomatic to use the [`Context::scope`](https://docs.rs/tracing-subscriber/0.2.5/tracing_subscriber/layer/struct.Context.html#method.scope) iterator here instead. `visit_spans` is essentially an older version of `scope` and might be deprecated eventually.

```suggestion
            let spans = ctx.scope()
                .map(visit_span)
                .collect::<Vec<_>>();
```

Reply:
Thanks, I switched to using `Context::scope`